File: frontend/services/simulation.ts
import { api } from "./api";

export interface AssetInfo {
  id: string;
  ticker: string;
  name: string;
  asset_class: string;
  description: string;
  data_available: boolean;
}

export interface PortfolioPoint {
  date: string;
  value: number;
  allocation: Record<string, number>;
}

export interface TradeRecord {
  date: string;
  asset: string;
  action: string;
  old_weight: number;
  new_weight: number;
  reason: string;
  trigger_trait: string;
  trait_value: number;
}

export interface EventAnnotation {
  date: string;
  label: string;
  impact: string;
}

export interface SimulationMetrics {
  total_return: number;
  annualized_return: number;
  max_drawdown: number;
  sharpe_ratio: number;
  volatility: number;
  best_month: number;
  worst_month: number;
  total_trades: number;
}

export interface SimulationResponse {
  portfolio_history: PortfolioPoint[];
  benchmark_history: PortfolioPoint[];
  trades: TradeRecord[];
  events: EventAnnotation[];
  metrics: SimulationMetrics;
  final_allocation: Record<string, number>;
  traits: Record<string, number>;
  persona_type: string;
}

export interface SimulationRequest {
  persona_vector?: number[];
  start_date?: string;
  end_date?: string;
  initial_capital?: number;
  asset_classes?: string[];
}

export async function getSimulationAssets(): Promise<AssetInfo[]> {
  const res = await api.get<AssetInfo[]>("/api/simulation/assets");
  return res.data;
}

export async function runSimulation(req: SimulationRequest): Promise<SimulationResponse> {
  const res = await api.post<SimulationResponse>("/api/simulation/run", req);
  return res.data;
}

export async function getMarketEvents(start?: string, end?: string): Promise<EventAnnotation[]> {
  const params: Record<string, string> = {};
  if (start) params.start = start;
  if (end) params.end = end;
  const res = await api.get<EventAnnotation[]>("/api/simulation/events", { params });
  return res.data;
}

File: backend/app/schemas/simulation.py
"""Pydantic v2 schemas for market simulation."""
from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class SimulationRequest(BaseModel):
    persona_vector: list[float] | None = None  # 16-dim; None = use user's portfolio persona
    start_date: date = date(2021, 1, 1)
    end_date: date = date(2026, 1, 1)
    initial_capital: float = 10000.0
    asset_classes: list[str] = ["stocks", "bonds", "gold", "bitcoin", "tech"]


class PortfolioPoint(BaseModel):
    date: str  # ISO date "YYYY-MM-DD"
    value: float
    allocation: dict[str, float]  # asset_class -> weight


class TradeRecord(BaseModel):
    date: str
    asset: str
    action: str  # "BUY" | "SELL" | "HOLD"
    old_weight: float
    new_weight: float
    reason: str  # human-readable
    trigger_trait: str  # which trait drove this
    trait_value: float


class EventAnnotation(BaseModel):
    date: str
    label: str
    impact: str  # "positive" | "negative" | "neutral"


class SimulationMetrics(BaseModel):
    total_return: float        # e.g. 0.45 = 45%
    annualized_return: float
    max_drawdown: float        # negative e.g. -0.25
    sharpe_ratio: float
    volatility: float          # annualized
    best_month: float
    worst_month: float
    total_trades: int


class SimulationResponse(BaseModel):
    portfolio_history: list[PortfolioPoint]
    benchmark_history: list[PortfolioPoint]  # SPY buy-and-hold
    trades: list[TradeRecord]
    events: list[EventAnnotation]
    metrics: SimulationMetrics
    final_allocation: dict[str, float]
    traits: dict[str, float]
    persona_type: str  # human label: "Conservative", "Aggressive", "FOMO Trader", etc.


class AssetInfo(BaseModel):
    id: str
    ticker: str
    name: str
    asset_class: str
    description: str
    data_available: bool
